[16253] MongoDB Storage Adapter Recipe README and Refinements. (#491)
ChartIQ/finsemble-seed: src/adapters/README.md

	 * @param {string} params.topic A topic under which the data should be stored.
	 * @param {string} params.key The key whose value is being set.
	 * @param {function} cb Callback to be invoked upon completion.
	 */
	get = function(params, cb) {
		
	}

	/**
	 * Save method.
	 * @param {object} params
	 * @param {string} params.topic A topic under which the data should be stored.
	 * @param {string} params.key The key whose value is being set.
	 * @param {any} params.value The value being saved.
	 * @param {function} cb Callback to be invoked upon save completion.
	 */
	save = function(params, cb) {
		
	}

    /**
	 * Returns all keys stored in MongoDB.
	 *
	 * @param {object} params
	 * @param {function} cb
	 */
	keys = function(params, cb) {

	}

	/**
	 * Delete method.
	 * @param {object} params
	 * @param {string} params.topic A topic under which the data should be stored.
	 * @param {string} params.key The key whose value is being deleted.
	 * @param {function} cb Callback to be invoked upon completion.
	 */
	delete = function(params, cb) {
		
	}
}

let mongoAdapter = new MongoStorageAdapter("MongoStorageAdapter")
module.exports = mongoAdapter; 

```

> Because keys are a multipart string, it's useful to have a helper function to construct what we'll refer to as the _combined key_. The helper `getCombinedKey` is provided for you in the baseStorage model within Finsemble; by extending our adapter from BaseStorage, it will inherit this helper. You should also include the Logger client for debugging purposes.

Reviewer: ```suggestion
The Storage service in Finsemble provides multiple variables that can be used to address stored data, however, in our MongoDB example, we're going to work with simple key:value pairs. Hence, keys are a multipart strings and therefore it's useful to have a helper function to construct what we'll refer to as the _combined key_. The helper `getCombinedKey` is provided for you in the baseStorage model within Finsemble; by extending our adapter from BaseStorage, it will inherit this helper. You should also include the Logger client for debugging purposes.

Note that you do not have to use the `getCombinedKey` helper, particularly if you intend to store data in a multi-column format using all of the variables provided.
```

            cb(err, { status: "failed" });
        });
};

```

Again, note the explicit invocation of the callback functions.

### `delete`

The **delete** method is similar to **save**:

```js
delete = function(params, cb) {
    let combinedKey = this.getCombinedKey(this, params);

    Logger.system.debug("Mongo.delete for key=" + combinedKey);

    fetch(`${this.ROUTES.DELETE}`, {
        method: "DELETE",
        body: JSON.stringify({ key: combinedKey }),
        headers: {
            'Content-Type': 'application/json'
        }
    })
    .then(data => {
        cb(null, { status: "success" });
    })      
    .catch(err => {
        Logger.system.error("Mongo.delete key=" + combinedKey + ", Error", err);
        cb(err, { status: "failed" });
    });
};

```

Your storage adapter might have some additional needs, such as a `clear` function to remove all data associated with a user or a `cacheClear` function to clear the database completely. For now, we're ready to test our adapter. These methods should at least be be stubbed out to prevent errors from Finsemble calling these and returning errors.

### Completed Adapter Example

> ```js

Reviewer: I'd remove this and refer to the file in the repository by URL/reference.

Reviewer: It ends up pushing the install instructions way down...

	 * particular user.
	 */
	clearCache = function(params, cb) {
		// Implementation here
	};

	/**
	 * Wipes the storage container.
	 * @param {function} cb
	 */
	empty = function(cb) {
		// Implementation here
	};
}

let mongoAdapter = new MongoStorageAdapter("MongoStorageAdapter")
module.exports = mongoAdapter; // Allows us to get access to the uninitialized object

```

## Finsemble Setup

1. Open _configs/application/config.json_ and locate the `storage` object: 
    ```js
	"storage": {
            "defaultStorage": "MongoStorageAdapter",
            "topicToDataStoreAdapters": {
                "finsemble": "MongoStorageAdapter",
                "finsemble.workspace": "MongoStorageAdapter",
                "finsemble.workspace.cache": "IndexedDBAdapter"
            },
            "dataStoreAdapters": {
                "LocalStorageAdapter": "$applicationRoot/adapters/localStorageAdapter.js",
                "IndexedDBAdapter": "$applicationRoot/adapters/indexedDBAdapter.js",
                "MongoStorageAdapter": "$applicationRoot/adapters/MongoStorageAdapter.js"
            }
        }
    ```

1. Add a `dataStoreAdapter` as shown above.

> 1. Register your new adapter as `defaultStorage` and for the  `finsemble` and `finsemble.workspace` topics. Note that `finsemble.workspace.cache` is a _high frequency_ topic (e.g., it's read and written every time a component is opened, dragged, resized, etc.) and thus should be maintained as a local storage adapter and _not_ saved to an external datastore.

Reviewer: ```suggestion
1. Register your new adapter as `defaultStorage` so that it is used for any unspecified storage topics.
1. Map the  `finsemble` and `finsemble.workspace` topics to your new adapter in `topicToDataStoreAdapters`. Note that `finsemble.workspace.cache` is a _high frequency_ topic (e.g., it's read and written every time a component is opened, dragged, resized, etc.) and thus should be maintained as a local storage adapter and _not_ saved to an external datastore.
```

> 1. Add your storage adapter to _build/webpack/webpack.adapters.entries.json_.

Reviewer: ```suggestion
1. Add your storage adapter to _build/webpack/webpack.adapters.entries.json_ to ensure it is built by webpack in the Finsemble seed project.
```
